### ISSUE TITLE
legacyTagKey does not maintain the legacy schema layout

### ISSUE THREAD

**Issue by sebastian-alfers** [akka/akka-persistence-jdbc#825]
[This](https://github.com/akka/akka-persistence-jdbc/commit/d4d942ec7b217606ce448178392daabcda1bf9a6) contribution introduced a new column `persistence_id` in the `event_tag` table. Also, a new flag `legacyTagKey` got introduced which was intended default to the backward compatibility code path that works without the new column.

However, looking at [this](https://github.com/akka/akka-persistence-jdbc/blob/main/core/src/main/scala/akka/persistence/jdbc/journal/dao/JournalQueries.scala#L42-L58) section shows that regardless of the value of `legacyTagKey` a `persistenceId` gets written. In the default case (`legacyTagKey=trure`) there is no such column.

**Comment by sebastian-alfers** [comment]
Looking through the implementation, I think that the migration docs were just not right. Looking at [this comment](https://github.com/akka/akka-persistence-jdbc/blob/main/core/src/test/scala/akka/persistence/jdbc/query/EventsByTagMigrationTest.scala#L191-L192), it gives the impression that the schema as actually intended to be changed (two columns added) without taking advantage of the improvements (which is `legacyTagKey=true`). So one immediate solution would be to run the [first part of the migration script](https://github.com/akka/akka-persistence-jdbc/blob/main/core/src/main/resources/schema/postgres/postgres-event-tag-migration.sql#L3-L5), e.g. for postgres:

```
ALTER TABLE public.event_tag
    ADD persistence_id  VARCHAR(255),
    ADD sequence_number BIGINT;
```

I assume this works since the query part does truly maintain the legacy schema layout by only looking at the `event_id` column.

Waiting for additional feedback if this issue just ends up just needing to fix the migration docs, or if we need a fix in the implementation to not insert into the two additional columns in the legacy-tag-key mode.

**Comment by johanandren** [comment]
From talking to @patriknw and @octonato the intent was to not require the schema change.

Could still be that it is to tricky to get working to be worth it, but we should at least try to get it working with the old schema.

**Comment by sebastian-alfers** [comment]
I did not find a good way to have a runtime-migration between the old and new schema which would allow a non-downtime migration to the non-legacy tag approach.

I think we should revert the default code path and schema to what is was pre 5.4.0, and then add explicit docs to explain how to go to the non-legacy tag mode - which requires a downtime.

Alternative: add to the docs that adding the two columns is required - which feels bad in a non-major release.

**Comment by patriknw** [comment]
> and then add explicit docs to explain how to go to the non-legacy tag mode - which requires a downtime

I don't think it will require downtime. That was the whole intention of the effort in adding the config flags. I think it's already possible to do a multi-phase rolling update, but it requires that the two columns are added as the first thing while it's running the old version. I agree that it's unfortunate to require adding those columns.

**Comment by octonato** [comment]
Done in #832
